### PR TITLE
[SPARK-23525] [SQL] Support ALTER TABLE CHANGE COLUMN COMMENT for external hive table

### PR DESCRIPTION
**Pull request by jiangxb1987**
## What changes were proposed in this pull request?

The following query doesn't work as expected:
```
CREATE EXTERNAL TABLE ext_table(a STRING, b INT, c STRING) PARTITIONED BY (d STRING)
LOCATION 'sql/core/spark-warehouse/ext_table';
ALTER TABLE ext_table CHANGE a a STRING COMMENT "new comment";
DESC ext_table;
```
The comment of column `a` is not updated, that's because `HiveExternalCatalog.doAlterTable` ignores table schema changes. To fix the issue, we should call `doAlterTableDataSchema` instead of `doAlterTable`.

## How was this patch tested?

Updated `DDLSuite.testChangeColumn`.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 1]
```diff
@@ -1597,6 +1597,7 @@ abstract class DDLSuite extends QueryTest with SQLTestUtils {
     // Ensure that change column will preserve other metadata fields.
     sql("ALTER TABLE dbx.tab1 CHANGE COLUMN col1 col1 INT COMMENT 'this is col1'")
     assert(getMetadata("col1").getString("key") == "value")
+    assert(getMetadata("col1").getString("comment") == "this is col1")
```
Comment (jiangxb1987) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: I didn't verify the comment is correctly set to the column metadata, and that caused the issue. :(

**Comment by jiangxb1987** [comment]
cc @skliarpawlo @gatorsmile @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala [hunk 2]
```diff
@@ -314,8 +314,8 @@ case class AlterTableChangeColumnCommand(
     val resolver = sparkSession.sessionState.conf.resolver
     DDLUtils.verifyAlterTableType(catalog, table, isView = false)
 
-    // Find the origin column from schema by column name.
-    val originColumn = findColumnByName(table.schema, columnName, resolver)
+    // Find the origin column from dataSchema by column name.
+    val originColumn = findColumnByName(table.dataSchema, columnName, resolver)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: Only data schema? How about partitioning schema?

Comment (jiangxb1987) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: As described in the comment, we only support altering non-partition columns:
```
 A command to change the column for a table, only support changing the comment of a non-partition
 column for now.
```

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: Do we have a negative test case to cover that?

File: sql/core/src/test/resources/sql-tests/results/change-column.sql.out [hunk 3]
```diff
@@ -291,16 +291,25 @@ ALTER TABLE partition_table PARTITION (c = 1) CHANGE COLUMN a new_a INT
...
 
 -- !query 30
-DROP TABLE test_change
+ALTER TABLE partition_table CHANGE COLUMN c c INT COMMENT 'this is column C'
 -- !query 30 schema
 struct<>
 -- !query 30 output
-
+org.apache.spark.sql.AnalysisException
+Invalid column reference 'c', table data schema is 'StructType(StructField(a,IntegerType,true), StructField(b,StringType,true))';
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/results/change-column.sql.out]: We need a better error message here.

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/results/change-column.sql.out]: how about 
```
Can't find column `c` given table data columns [`a`, `b`]
```

**Comment by cloud-fan** [comment]
LGTM

**Comment by cloud-fan** [comment]
thanks, merging to master/2.3!

**Pull request closed without merging**